src/doc/unstable-book/src/language-features/macro-metavar-expr-concat.md

```diff
@@ -0,0 +1,154 @@
+# `macro_metavar_expr_concat`
+
+The tracking issue for this feature is: [#124225]
+
+------------------------
+
+
+`#![feature(macro_metavar_expr_concat)]` provides a more powerful alternative to [`concat_idents!`].
+
+> This feature is not to be confused with [`macro_metavar_expr`] or [`concat_idents`].
+
+> This is an experimental feature; it and its syntax will require a RFC before stabilization.
+
+
+### Overview
+
+`macro_rules!` macros cannot create new identifiers and use them in ident positions.
+A common use case is the need to create new structs or functions. The following cannot be done[^1]:
+
+```rust,compile_fail
+macro_rules! create_some_structs {
+  ($name:ident) => {
+      // Invalid syntax
+      struct First_$name;
+       // Also invalid syntax
+      struct Second_($name);
+      // Macros are not allowed in this position
+      // (This restriction is what makes `concat_idents!` useless)
+      struct concat_idents!(Third_, $name);
+  }
+}
+# create_some_structs!(Thing);
+```
+
+`#![feature(macro_metavar_expr_concat)]` provides the `concat` metavariable to concatenate idents in ident position:
+
+```rust
+#![feature(macro_metavar_expr_concat)]
+# #![allow(non_camel_case_types, dead_code)]
+
+macro_rules! create_some_structs {
+  ($name:ident) => {
+      struct ${ concat(First_, $name) };
+      struct ${ concat(Second_, $name) };
+      struct ${ concat(Third_, $name) };
+  }
+}
+
+create_some_structs!(Thing);
+```
+
+This macro invocation expands to:
+
+```rust
+# #![allow(non_camel_case_types, dead_code)]
+struct First_Thing;
+struct Second_Thing;
+struct Third_Thing;
+```
+
+### Syntax
+
+This feature builds upon the metavariable expression syntax `${ .. }` as specified in [RFC 3086] ([`macro_metavar_expr`]).
+ `concat` is available like `${ concat(items) }`, where `items` is a comma separated sequence of idents and/or string literals.
```

Comment:
`s/string literals/literals`, I think we want to support numeric literals as well (or maybe already do?)

Reply:
numerics are not supported: https://play.rust-lang.org/?version=nightly&mode=release&edition=2024&gist=db1dbedf669a4061ff6ebceccf9cd7ad
```rust
#![feature(macro_metavar_expr_concat)]
macro_rules! cursed_add {
  ($a:literal, $b:literal) => {
      ${ concat($a, $b) }
  }
}

fn main() {
    let x = cursed_add!(42, 73);
    dbg!(x);
}
```
```
error: metavariables of `${concat(..)}` must be of type `ident`, `literal` or `tt`
 --> src/main.rs:4:18
  |
4 |       ${ concat($a, $b) }
  |                  ^
  |
  = note: currently only string literals are supported
```